### PR TITLE
[Python] Support loading of TF models with saved weights

### LINKED ISSUE #25366: [Feature Request]: Implement a native Tensorflow model hander
### What would you like to happen?

Currently we use the tensorflow model handler in tfx_bsl which has some constraints listed in https://github.com/apache/beam/issues/23584. Implement a new model handler with support for more types like numpy, tensors, etc.

- [x] Load an entire saved model 
- [x] Load model with saved weights

### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by riteshghorse**
Adds support for loading tensorflow model from saved weights given a function to create the model.
Follow up to #25368, Closes #25366 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by riteshghorse** [comment]
`Python_PVR_Flink` failed because of licensing issue. Bruno has a PR at https://github.com/apache/beam/pull/25515

**Comment by riteshghorse** [comment]
Run Python_PVR_Flink

**Comment by riteshghorse** [comment]
Run Python_PVR_Flink PreCommit

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

**Comment by damccorm** [comment]
I think you need to pull in master to avoid postcommit failures (fixed by https://github.com/apache/beam/pull/25446)

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

**Comment by riteshghorse** [comment]
Run Python_Dataframes PreCommit

**Comment by riteshghorse** [comment]
Run Whitespace PreCommit

File: sdks/python/tox.ini
Comment (damccorm) [inline sdks/python/tox.ini:347]: Hm, looks like all tf tests are currently being skipped - https://ci-beam.apache.org/job/beam_PostCommit_Python38_PR/697/testReport/apache_beam.ml.inference.tensorflow_inference_it_test/TensorflowInference/

I think to get these to run correctly, we should add `tensorflow` to our test dependencies - https://github.com/apache/beam/blob/f358fcdc8c8a844e1bdfc8ca8f6fe886734c8df9/sdks/python/setup.py#L288

It actually looks like we have the same problem with pytorch and onnx, but that could be a follow up PR (from you or someone else)

Comment (riteshghorse) [reply sdks/python/tox.ini:347]: Oooh.. Thanks for catching that

Comment (riteshghorse) [reply sdks/python/tox.ini:347]: Created https://github.com/apache/beam/issues/25547 to track pytorch and onnx

**Comment by riteshghorse** [comment]
Run Python 3.8 PostCommit

File: sdks/python/tox.ini
Comment (damccorm) [reply sdks/python/tox.ini:347]: Thanks - I have a draft PR for it (https://github.com/apache/beam/pull/25548)

File: sdks/python/setup.py
Comment (tvalentyn) [inline sdks/python/setup.py:289]: let's not depend on a large dependency like this by default. you can see how we deal with sklearn/pytorch in  similar circumstances.

Comment (AnandInguva) [reply sdks/python/setup.py:289]: +1 to @tvalentyn.

Comment (AnandInguva) [reply sdks/python/setup.py:289]: Commented at https://github.com/apache/beam/issues/25547#issuecomment-1435716734. 

We can follow the approach of sklearn/torch/tensorRT gradle tests and not add very large ML dependencies to the setup.py

Comment (damccorm) [reply sdks/python/setup.py:289]: Ah, I see - so this was running on 3.7 and 3.9, just not 3.8 because of the difference in dependencies here - https://github.com/apache/beam/blob/921bc7b94708391eb7d980cb8d9ed35c80d3aebb/build.gradle.kts#L457

Thanks for clarifying. I'm now +1 on removing this and manually triggering Python 3.7 postcommits, sorry for my confusion @riteshghorse 

---------------------------

A couple of follow up questions for @AnandInguva and @tvalentyn :
1) Is there any reason we only run the dataflow versions on 3.8? Since TensorRT is on Dataflow, we now need to run both the 3.7 and 3.8 postcommits to fully exercise our inference code base
2) Is there any reason we're not running most of our inference postcommits on dataflow (as well as the direct runner)? That's probably where most of our usage is today

**Comment by riteshghorse** [comment]
Run Python 3.7 PostCommit

**Comment by riteshghorse** [comment]
Run Python Unit Tests

**Comment by riteshghorse** [comment]
Run Python 3.9 PostCommit

**Comment by damccorm** [comment]
3.7 has passing tensorflow tests. I'll wait until all active suites complete and merge

image

**Review by damccorm — APPROVED** [review]

**Comment by damccorm** [comment]
Run Python_Coverage PreCommit

**Comment by damccorm** [comment]
(all other checks have passed, they're just not statusing)

**Comment by riteshghorse** [comment]
All checks passed. Merging!

**Pull request merged**
